llvm/test/Transforms/SimplifyCFG/X86/hoist-load-store-with-cf.ll

```diff
@@ -0,0 +1,460 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```

Comment:
Can you add some tests with gep?
```
define i32 @cload(ptr %x)  {
entry:
  %tobool.not = icmp eq ptr %x, null
  br i1 %tobool.not, label %cond.end, label %cond.true

cond.true:                                        ; preds = %entry
  %arrayidx = getelementptr inbounds i8, ptr %x, i64 16
  %0 = load i32, ptr %arrayidx, align 4, !tbaa !9
  br label %cond.end

cond.end:                                         ; preds = %entry, %cond.true
  %cond = phi i32 [ %0, %cond.true ], [ 0, %entry ]
  ret i32 %cond
}
```

Reply:
Test (`load_from_gep`) was added.  `arrayidx` is defined in the successor and not a result of load, and we don't hoist the `load` to avoid recursively check the operands.

Reply:
> You'd like me to support the operand of load to be a GEP in this PR? It's doable but just increases the complexity a bit.

No. We can improve `load_from_gep` in follow-up patches :)